sympy/calculus/util.py

```diff
@@ -476,7 +476,9 @@ def _check(orig_f, period):
             for index, g in enumerate(reversed(g_s)):
                 start_index = num_of_gs - 1 - index
                 g = compogen(g_s[start_index:], symbol)
-                if g != orig_f and g != f: # Fix for issue 12620
+                # Fix for issue 12620
+                if decompogen(g, symbol) != decompogen(orig_f, symbol) \
```

Comment:
Would it be advantageous to factor out the orig_g calculation so that doesn't have to be repeatedly calculated for every g? Also, you are comparing `decompogen(g, symbol)` (an element-ish thing of `g_s`) with `decompogen(f, symbol)`...which *is* `g_s`. Will those two every actually be equal? And, again, is it necessary to calculate `decompogen(g, symbol)` twice to do this test?

Since g of the `for` loop is never used, please re-write that as `for index in range(len(g_s)):`...but I wonder if some edit went awry since why is the `g_s` being reversed if the `g` is never used? /!\ CAUTION

Reply:
Comparing `decompogen(g, symbol)` with `decompogen(f, symbol)`, makes sure that the function does not enter infinite recursion, because even though the decomposition of two functions might be the same, the functions might be different, as was happening in the test case added. Because the `periodicity` uses recursion to break down a function into functions for which the periodicity can be directly calculated, this causes the infinite recursion.

Removing `g` from the `for` loop and rewriting as suggested does seems less complicated. There seems to be no problem, as both the loops do the exact same thing.

If by removing the `orig_f` calculation, you mean the comparison on L479, I don't think it should be removed. `f` is getting reassigned in some places, and so comparison to `orig_f` should be done as well.

I shall correct the part where `decompogen(g, symbol)` is being computed twice.